server/app/models.py

```diff
@@ -658,15 +660,50 @@ def mark_as_final(self):
             final = FinalSubmission.query(
                 FinalSubmission.assignment==assignment,
                 FinalSubmission.submitter==submitter).get()
+        return final
+
+    def mark_as_final(self):
+        """Create or update a final submission."""
+        final = self.get_final()
+
         if final:
             final.submitter = self.submitter
             final.submission = self.key
         else:
-            final = FinalSubmission(assignment=assignment, submission=self.key)
+            group = self.submitter.get().get_group(self.assignment)
+            final = FinalSubmission(
+                assignment=self.assignment, submission=self.key)
             if group:
                 final.group = group.key
         final.put()
 
+    def resubmit(self, user_key):
```

Comment:
This function is pretty simple... I feel like there's a lot of stuff we have to do to make the submission process itself do what we want, which is actually (I believe) more complicated than this. But, we could use that if you want.